web/src/app/rotations/RotationForm.js

```diff
@@ -1,54 +1,94 @@
+
 import { FormContainer, FormField } from '../forms'
 import { TimeZoneSelect } from '../selection'
-import { TextField, Grid, MenuItem } from '@material-ui/core'
-import { startCase } from 'lodash'
-import { DateTime, Info } from 'luxon'
-import { ISOTimePicker } from '../util/ISOPickers'
+import { ISODateTimePicker } from '../util/ISOPickers'
+import NumberField from '../util/NumberField'
+import Spinner from '../loading/components/Spinner'
+
+const query = gql`
+  query nextRotationHandoffTimes($input: NextRotationHandoffTimesInput) {
+    nextRotationHandoffTimes(input: $input)
+  }
+`
 
 const rotationTypes = ['hourly', 'daily', 'weekly']
 
-export default function RotationForm(props) {
-  function dayOfWeek() {
-    const { start, timeZone } = props.value
-    return DateTime.fromISO(start, { zone: timeZone }).weekday
-  }
+const useStyles = makeStyles({
+  listNone: {
+    listStyle: 'none',
+  },
+  bolder: {
+    fontWeight: 'bolder',
+  },
+  flex: {
+    display: 'flex',
+  },
+  height7rem: {
+    height: '7rem',
+  },
+  noSpacing: {
+    margin: 0,
+    padding: 0,
+  },
+})
 
-  function setDayOfWeek(weekday) {
-    const { start, timeZone, ...other } = props.value
-    props.onChange({
-      ...other,
-      timeZone,
-      start: DateTime.fromISO(start, { zone: timeZone })
-        .set({ weekday })
-        .toISO(),
-    })
+// getHours converts a count and one of ['hourly', 'daily', 'weekly']
+// into length in hours e.g. (2, daily) => 48
+function getHours(count, unit) {
+  const lookup = {
+    hourly: 1,
+    daily: 24,
+    weekly: 24 * 7,
   }
+  return lookup[unit] * count
+}
 
-  function renderDayOfWeekField() {
-    return (
-      <Grid item xs={12}>
-        <TextField
-          fullWidth
-          select
-          required
-          label='Day of Week'
-          name='dayOfWeek'
-          value={dayOfWeek()}
-          onChange={(e) => setDayOfWeek(e.target.value)}
-        >
-          {Info.weekdaysFormat('long').map((day, idx) => {
-            return (
-              <MenuItem key={day} value={idx + 1}>
-                {day}
-              </MenuItem>
-            )
-          })}
-        </TextField>
-      </Grid>
-    )
-  }
+export default function RotationForm(props) {
+  const { value } = props
+  const classes = useStyles()
+  const localZone = DateTime.local().zone.name
+  const [configInZone, setConfigInZone] = useState(false)
+  const configZone = configInZone ? value.timeZone : 'local'
+
+  const [minStart, maxStart] = useMemo(() => [
+    DateTime.local().minus({ year: 1 }),
+    DateTime.local().plus({ year: 1 }),
+  ])
+
+  const { data, loading, error } = useQuery(query, {
+    variables: {
+      input: {
+        start: value.start,
+        timeZone: value.timeZone,
+        hours: getHours(value.shiftLength, value.type),
+        count: 3,
+      },
+    },
+  })
+
+  const isCalculating = !data || loading
+
+  const nextHandoffs = isCalculating
```

Comment:
If you put this in a `useEffect` where we only update the array when new data comes in from `nextRotationHandoffTimes`, we should see the flicker go away

Reply:
We moved away from that to keep from duplicating/double-caching data. We also don't want to show stale data while fetching with no indication of it being invalid (e.g., delayed network request).

This would be a good use-case for that early spinner design we had -- show the old/no spinner if the new comes within < 100ms otherwise spin for at least 1 second etc.. Normal use will be to update the handoff zero or one time and a flicker on that update is probably an okay compromise.

Reply:
Gotcha, that makes sense. We could alternatively find and replace only the characters that change, rather than the whole container. I'm not sure how it would play out, but could potentially work in our favor.

Reply:
Not sure I know what you mean, React already reconciles the virtual DOM after render, and only updates the changed bits. The flicker is because we're rendering times, then the spinner, then the new times again in a short timeframe.

Maybe we can look into tools/methods of debouncing render for these situations? Maybe we should keep that out of scope of this PR though, I don't think the flicker on change detracts enough from the use to block it.